pkg/registrar/bastion.go

```diff
@@ -0,0 +1,138 @@
+package registrar
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	"github.com/giantswarm/microerror"
+	"github.com/go-logr/logr"
+	clouddns "google.golang.org/api/dns/v1"
+	capg "sigs.k8s.io/cluster-api-provider-gcp/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+)
+
+//counterfeiter:generate . BastionsClient
+type BastionsClient interface {
+	GetBastionIPList(ctx context.Context, cluster *capg.GCPCluster) ([]string, error)
+	AddFinalizerToBastions(ctx context.Context, cluster *capg.GCPCluster) error
+	RemoveFinalizerFromBastions(ctx context.Context, cluster *capg.GCPCluster) error
+}
+
+type Bastion struct {
+	baseDomain     string
+	bastionsClient BastionsClient
+	dnsService     *clouddns.Service
+}
+
+func NewBastion(baseDomain string, bastionsClient BastionsClient, dnsService *clouddns.Service) *Bastion {
+	return &Bastion{
+		baseDomain:     baseDomain,
+		bastionsClient: bastionsClient,
+		dnsService:     dnsService,
+	}
+}
+
+func (r *Bastion) Register(ctx context.Context, cluster *capg.GCPCluster) error {
+	logger := r.getLogger(ctx)
+	err := r.bastionsClient.AddFinalizerToBastions(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	bastionIPList, err := r.bastionsClient.GetBastionIPList(ctx, cluster)
+	if err != nil {
+		return microerror.Mask(err)
+	}
+
+	for i, bastionIP := range bastionIPList {
+		bastionDomain := fmt.Sprintf("%s.%s.%s.", EndpointBastion(i+1), cluster.Name, r.baseDomain)
+		logger := logger.WithValues("record", bastionDomain)
+		logger.Info("Registering record")
+
+		record := &clouddns.ResourceRecordSet{
+			Name: bastionDomain,
+			Rrdatas: []string{
+				bastionIP,
+			},
+			Type: RecordA,
+		}
+		_, err = r.dnsService.ResourceRecordSets.Create(cluster.Spec.Project, cluster.Name, record).
+			Context(ctx).
+			Do()
+
+		if hasHttpCode(err, http.StatusConflict) {
+			// record exists, check if the IP matches
+			rr, err := r.dnsService.ResourceRecordSets.Get(cluster.Spec.Project, cluster.Name, bastionDomain, RecordA). //nolint:govet
+																	Context(ctx).
+																	Do()
+			if err != nil {
+				return microerror.Mask(err)
+			}
+
+			if len(rr.Rrdatas) != 1 || rr.Rrdatas[0] != bastionIP {
+				logger.Info("Bastion record exists but its not up to date. Updating record")
+
+				_, err = r.dnsService.ResourceRecordSets.Patch(cluster.Spec.Project, cluster.Name, bastionDomain, RecordA, record).
+					Context(ctx).
+					Do()
+				if err != nil {
+					return microerror.Mask(err)
+				}
+				logger.Info("Updated Bastion record", "ip", bastionIP)
+			} else {
+				logger.Info("Skipping. Record already exists")
+				continue
+			}
```

Comment:
A few things here:
1. This looks like it should be a function of its own
1. the `len(rr.Rrdatas) != 1` sounds like an error. If we have a record with 0 or 2 or more Rrdatas, that sounds like it somehow got messed up. We only expect it to have one IP in there.
1. And most importantly, I think this entire thing relies on the bastionIPs in the loop to always be returned in the same order, which I'm pretty sure we have no guarantee on. This means that the domains probably won't be stable, so at one point `bastion1` will point to one IP, but after an update to the Cluster CR, it might point to another
1. Also if the number of IPs is less than it was originally, there's nothing that can clean up the delta

Reply:
1 moved to function
2 it's more like a safe code so it doesn't panic when the record is empty for unknow reason ( like customer or manual changes)
3) the order does not matter if the order matter then the IP will be updated but in the end, we don't care if bastionA has IP 1.2.3.4 and bastionB has IP 1.2.3.5 or if they are reversed. we still only care we can ssh either to BastionA or bastionB

the Ip changes on update anyway

4) now during deletion we delete all bastion records